File: elixir/scl_parser/lib/scl/serializer.ex
defmodule SCL.Serializer do
  @moduledoc false

  alias SCL.Errors.ParseError

  @indent 4

  def serialize(map) when is_map(map) do
    do_serialize(map, 0) <> "\n"
  end

  defp do_serialize(map, level) do
    indent = String.duplicate(" ", @indent * level)
    map
    |> Enum.map(fn {k, v} ->
      key = to_string(k)
      indent <> key <> " :: " <> serialize_value(v, level)
    end)
    |> Enum.join("\n")
  end

  # bool
  defp serialize_value(v, _level) when is_boolean(v) do
    "bool { " <> (if v, do: "true", else: "false") <> " }"
  end

  # int
  defp serialize_value(v, _level) when is_integer(v) do
    "num { " <> Integer.to_string(v) <> " }"
  end

  # float
  defp serialize_value(v, _level) when is_float(v) do
    "fl { " <> :erlang.float_to_binary(v, [:compact]) <> " }"
  end

  # string
  defp serialize_value(v, level) when is_binary(v) do
    if String.contains?(v, "\n") do
      indent = String.duplicate(" ", @indent * level)
      "ml {\n" <> indent <> "    '" <> v <> "'\n" <> indent <> "}"
    else
      escaped =
        v
        |> String.replace("\\", "\\\\")
        |> String.replace("\"", "\\\"")
      "str { \"" <> escaped <> "\" }"
    end
  end

  # map
  defp serialize_value(v, level) when is_map(v) do
    "class {\n" <> do_serialize(v, level + 1) <> "\n" <> String.duplicate(" ", @indent * level) <> "}"
  end

  # list
  defp serialize_value(v, _level) when is_list(v) do
    case v do
      [] -> "list(str) { }"
      _ ->
        first = hd(v)
        cond do
          is_boolean(first) and Enum.all?(v, &is_boolean/1) ->
            "list(bool) { " <> Enum.map_join(v, ", ", fn x -> if x, do: "true", else: "false" end) <> " }"
          is_integer(first) and Enum.all?(v, &is_integer/1) ->
            "list(num) { " <> Enum.map_join(v, ", ", &Integer.to_string/1) <> " }"
          (is_float(first) or is_integer(first)) and Enum.all?(v, &(is_float(&1) or is_integer(&1))) ->
            floats = Enum.map(v, fn x -> if is_integer(x), do: Integer.to_string(x), else: :erlang.float_to_binary(x, [:compact]) end)
            "list(fl) { " <> Enum.join(floats, ", ") <> " }"
          is_binary(first) and Enum.all?(v, &is_binary/1) ->
            escaped = Enum.map(v, fn s -> s |> String.replace("\\", "\\\\") |> String.replace("\"", "\\\"") end)
            "list(str) { " <> Enum.map_join(escaped, ", ", fn s -> "\"" <> s <> "\"" end) <> " }"
          true ->
            raise ParseError, message: "Unsupported list element: #{inspect(first)}"
        end
    end
  end

  # fallback
  defp serialize_value(v, _level) do
    cond do
      is_boolean(v) -> serialize_value(v, 0)
      is_integer(v) -> "dynamic { " <> Integer.to_string(v) <> " }"
      is_float(v) -> "dynamic { " <> :erlang.float_to_binary(v, [:compact]) <> " }"
      is_binary(v) ->
        escaped = v |> String.replace("\\", "\\\\") |> String.replace("\"", "\\\"")
        "dynamic { \"" <> escaped <> "\" }"
      true ->
        raise ParseError, message: "Unsupported type: #{inspect(v)}"
    end
  end
end

File: ruby/lib/scl/serializer.rb
require_relative "errors"

module SCL
  class Serializer
    def initialize(indent: 4)
      @indent = indent
    end

    def serialize(hash)
      raise ParseError, "Expected Hash" unless hash.is_a?(Hash)
      (serialize_obj(hash, 0) + "\n")
    end

    private

    def serialize_obj(hash, level)
      indent = " " * (@indent * level)
      hash.map do |k, v|
        key = k.to_s
        indent + key + " :: " + serialize_value(v, level)
      end.join("\n")
    end

    def serialize_value(v, level)
      case v
      when TrueClass, FalseClass
        "bool { #{v ? "true" : "false"} }"
      when Integer
        "num { #{v} }"
      when Float
        "fl { #{v} }"
      when String
        if v.include?("\n")
          indent = " " * (@indent * level)
          "ml {\n#{indent}    '#{v}'\n#{indent}}"
        else
          escaped = v.gsub("\\", "\\\\").gsub("\"", "\\\"")
          "str { \"#{escaped}\" }"
        end
      when Hash
        "class {\n" + serialize_obj(v, level + 1) + "\n" + (" " * (@indent * level)) + "}"
      when Array
        serialize_list(v)
      else
        serialize_dynamic(v, level)
      end
    end

    def serialize_list(arr)
      return "list(str) { }" if arr.empty?
      first = arr.first
      if arr.all? { |x| x.is_a?(TrueClass) || x.is_a?(FalseClass) }
        items = arr.map { |x| x ? "true" : "false" }.join(", ")
        "list(bool) { #{items} }"
      elsif arr.all? { |x| x.is_a?(Integer) }
        items = arr.map(&:to_s).join(", ")
        "list(num) { #{items} }"
      elsif arr.all? { |x| x.is_a?(Float) || x.is_a?(Integer) }
        items = arr.map { |x| x.is_a?(Integer) ? x.to_s : x.to_s }.join(", ")
        "list(fl) { #{items} }"
      elsif arr.all? { |x| x.is_a?(String) }
        items = arr.map { |s| "\"#{s.gsub("\\", "\\\\").gsub("\"", "\\\"")}\"" }.join(", ")
        "list(str) { #{items} }"
      else
        raise ParseError, "Unsupported list element type: #{first.class}"
      end
    end

    def serialize_dynamic(v, level)
      case v
      when TrueClass, FalseClass
        "dynamic { #{v ? "true" : "false"} }"
      when Integer
        "dynamic { #{v} }"
      when Float
        "dynamic { #{v} }"
      when String
        if v.include?("\n")
          indent = " " * (@indent * level)
          "ml {\n#{indent}    '#{v}'\n#{indent}}"
        else
          escaped = v.gsub("\\", "\\\\").gsub("\"", "\\\"")
          "dynamic { \"#{escaped}\" }"
        end
      else
        raise ParseError, "Unsupported value type: #{v.class}"
      end
    end
  end
end


